src/node/rpc/frontend.h

```diff
@@ -551,15 +545,20 @@ namespace ccf
 
       auto caller_id = endpoints.get_caller_id(tx, ctx->session->caller_cert);
 
-      if (
-        consensus != nullptr && consensus->type() == ConsensusType::BFT &&
-        (ctx->execute_on_node || consensus->is_primary()))
+      auto endpoint = endpoints.find_endpoint(tx, *ctx);
+
+      const bool is_bft =
+        consensus != nullptr && consensus->type() == ConsensusType::BFT;
+      const bool is_local =
+        endpoint != nullptr && endpoint->properties.execute_locally;
+      const bool should_bft_distribute = is_bft && !is_local &&
+        (ctx->execute_on_node || consensus->is_primary());
+
+      // TODO: This decision is based on several things read from the KV
+      // (cert->caller_id, request->is_local) which are true _now_ but may not
+      // be true when this is actually received/executed
```

Comment:
It's likely that we will need something like an endpoint definition digest which we capture upfront and re-verify at execution time in contexts like forwarding.

Reply:
Good point, I've added a mention of the plan for digest comparisons.